### PR TITLE
feat: add strict and custom thresholds support for relativeTime plugin

### LINKED ISSUE #830: How to get precise second(s) in `relativeTime` plugin?
**Describe the bug**
https://github.com/iamkun/dayjs/blob/dev/src/plugin/relativeTime/index.js#L8
All `seconds` displays `a few seconds`

**Expected behavior**
I want to display `1 second, 2 seconds...`

**Information**
 - Day.js Version: 1.8.22

### PR DESCRIPTION
**Pull request by JounQin**
close #830

### CODE REVIEW

File: src/plugin/relativeTime/index.js [hunk 1]
```diff
@@ -1,6 +1,36 @@
...
+
+const STRICT_THRESHOLDS = [
+  { l: 's', r: 1 },
+  { l: 'ss', r: 59, d: C.S },
+  { l: 'm', r: 1 },
+  { l: 'mm', r: 59, d: C.MIN },
+  { l: 'h', r: 1 },
+  { l: 'hh', r: 23, d: C.H },
+  { l: 'd', r: 1 },
+  { l: 'dd', r: 30, d: C.D },
```
Comment (JounQin) [inline src/plugin/relativeTime/index.js]: Should it be `29` or `30`?

File: src/utils.js [hunk 2]
```diff
@@ -45,6 +45,7 @@ const prettyUnit = (u) => {
 const isUndefined = s => s === undefined
 
 export default {
+  C, // for custom plugin
```
Comment (iamkun) [inline src/utils.js]: what's this export for?

File: src/plugin/relativeTime/index.js [hunk 3]
```diff
@@ -17,21 +47,23 @@ export default (o, c, d) => {
     y: 'a year',
     yy: '%d years'
   }
+  if (o.strict) {
+    Object.assign(d.en.relativeTime, {
```
Comment (iamkun) [inline src/plugin/relativeTime/index.js]: please avoid using es6 syntax `Object.assign`

File: src/utils.js (hunk 2 above)
Comment (JounQin) [reply src/utils.js]: Tests or user custom plugin.

https://github.com/iamkun/dayjs/pull/851/files#diff-f870cbe6ffaae01f79bb9242311d1f1aR133

If I'm going to write a custom plugin, I have to copy these constants manually before.

File: src/plugin/relativeTime/index.js (hunk 3 above)
Comment (JounQin) [reply src/plugin/relativeTime/index.js]: But `Object.keys` is used in [updateLocale](https://github.com/iamkun/dayjs/blob/dev/src/plugin/updateLocale/index.js#L6).

Comment (iamkun) [reply src/plugin/relativeTime/index.js]: `Object.keys` is not introduced in es6

Comment (JounQin) [reply src/plugin/relativeTime/index.js]: Ah, sorry about that, how about this `extend` util?

File: src/plugin/relativeTime/index.js [hunk 4]
```diff
@@ -17,21 +48,23 @@ export default (o, c, d) => {
     y: 'a year',
     yy: '%d years'
   }
+  if (o.strict) {
+    U.x(d.en.relativeTime, {
```
Comment (iamkun) [inline src/plugin/relativeTime/index.js]: can we just do 
```
d.en.relativeTime = {
  // and copy all the relevant keys
}
```

Comment (JounQin) [reply src/plugin/relativeTime/index.js]: I don't think it's a good pattern personally, but you'd like to do this, that's fine to me.

Comment (iamkun) [reply src/plugin/relativeTime/index.js]: Or, do we really have to update this? Why not just leave it to our user themself?

Comment (JounQin) [reply src/plugin/relativeTime/index.js]: Would it be much easier to use for users?

**Comment by iamkun** [comment]
After carefully thinking about this PR, it seems it is enough for us to just add a `thresholds` option. And leave the rest to our end user. 

How do you think?

**Comment by JounQin** [comment]
I think a `strict` preset would be helpful commonly, actually I don't understand why `moment` choose the loose way by default which is counterintuitive to me personally.

**Comment by iamkun** [comment]
Here's my concern, If we added `strict` mode, should we update all other locales `relativeTime` part as well? Otherwise, this might be meanless for some locale.

**Comment by iamkun** [comment]
However, if we just add a custom threshold, and leave the configuration to our user, it will be much better. All our user have to do is to use `UpdateLocale` to update `relativeTime` locale, to keep the sync with their custom thresholds.

**Comment by JounQin** [comment]
> Otherwise, this might be meanless for some locale.

Yes, users have to update locales except built-in English.
But using `updateLocale` seems always required for other locales.
With `strict` option, we keep built-in locale simple.

Another idea would be supporting `strict` `relativeTime` option in locale file, but maybe it's not worth to do it.

```js
module.exports = {
  relativeTime: {
    s: 'a few seconds',
    _s: '%d second', // strict text
    ss: '%d seconds' // fallback to `ss` if `_ss` is not present in `strict` mode
  }
}
```

File: src/plugin/relativeTime/index.js [hunk 5]
```diff
@@ -44,10 +84,10 @@ export default (o, c, d) => {
           ? d(input).diff(instance, t.d, true)
           : instance.diff(input, t.d, true)
       }
-      const abs = Math.round(Math.abs(result))
+      const abs = Math[o.strict ? 'floor' : 'round'](Math.abs(result))
```
Comment (JounQin) [inline src/plugin/relativeTime/index.js]: Another difference between two modes is here, `0.1s` would display `0 second` in `strict` mode.

Comment (iamkun) [reply src/plugin/relativeTime/index.js]: we can leave this to Additional token processing https://day.js.org/docs/en/customization/relative-time#additional-token-processing

**Comment by iamkun** [comment]
As for now, how about adding the `thresholds` option in this coming release only. And waiting for issues reply to decide whether to introduce `strict` mode or not.

File: types/index.d.ts [hunk 6]
```diff
@@ -96,9 +96,9 @@ declare namespace dayjs {
     locale(preset: string | ILocale, object?: Partial<ILocale>): Dayjs
   }
 
-  export type PluginFunc = (option: any, c: typeof Dayjs, d: typeof dayjs) => void
+  export type PluginFunc<T = unknown> = (option: T, c: typeof Dayjs, d: typeof dayjs) => void
 
-  export function extend(plugin: PluginFunc, option?: any): Dayjs
+  export function extend<T = unknown>(plugin: PluginFunc<T>, option?: T): Dayjs
```
Comment (iamkun) [inline types/index.d.ts]: I don't know much about TypeScript. What's this change about, please?

Comment (JounQin) [reply types/index.d.ts]: Stricter plugin options.

```ts
dayjs.extend(relativeTime, {
  Strict: true // error
})

dayjs.extend(relativeTime, {
  strict: true // pass
})
```

File: src/plugin/relativeTime/index.js (hunk 5 above)
Comment (JounQin) [reply src/plugin/relativeTime/index.js]: Then we have to write a few methods for `ss`, `mm`, `hh`, `dd`, etc.?
That seems terrible to me.

**Comment by JounQin** [comment]
> As for now, how about adding the `thresholds` option in this coming release only. And waiting for issues reply to decide whether to introduce `strict` mode or not.

I don't think it is feasible about "waiting for issues reply" personally.
There is only a few users care about it, maybe. 😂

**Comment by iamkun** [comment]
https://momentjs.com/docs/#/customization/relative-time-threshold/
https://momentjs.com/docs/#/customization/relative-time-rounding/

moment has these two options.  We can do this too to suit your needs.
```js
{
threshold: {},
rounding: Math.floor
}
```

**Comment by JounQin** [comment]
@iamkun Updated according to your last comment.

File: src/plugin/relativeTime/index.js [hunk 7]
```diff
@@ -1,11 +1,13 @@
 import * as C from '../../constant'
 
 export default (o, c, d) => {
+  o = o || {}
   const proto = c.prototype
   d.en.relativeTime = {
     future: 'in %s',
     past: '%s ago',
     s: 'a few seconds',
+    ss: '%d seconds',
```
Comment (iamkun) [inline src/plugin/relativeTime/index.js]: seems we don't need this key.

File: src/utils.js (hunk 2 above)
Comment (iamkun) [reply src/utils.js]: We better not doing this. The constant object will increase a huge bundle size.

File: src/plugin/relativeTime/index.js (hunk 7 above)
Comment (JounQin) [reply src/plugin/relativeTime/index.js]: `ss` is built-in supported by `moment`, and if we don't set it here, if user passes a custom `ss` threshold, its default format would be `undefined`.

Comment (iamkun) [reply src/plugin/relativeTime/index.js]: we can leave this to our user, I think. Same reason, we can't update the English locale only.

File: src/utils.js (hunk 2 above)
Comment (JounQin) [reply src/utils.js]: OK, I found `constant.js` is preserved in `esm` folder, but unavailable in root dir.

**Comment by iamkun** [comment]
That's great! Thanks.

File: types/index.d.ts [hunk 8]
```diff
@@ -96,9 +96,9 @@ declare namespace dayjs {
     locale(preset: string | ILocale, object?: Partial<ILocale>): Dayjs
   }
 
-  export type PluginFunc = (option: any, c: typeof Dayjs, d: typeof dayjs) => void
+  export type PluginFunc<T = unknown> = (option: T, c: typeof Dayjs, d: typeof dayjs) => void
```
Comment (iamkun) [inline types/index.d.ts]: One last question, what's this <T = unknown> for?

Comment (JounQin) [reply types/index.d.ts]: For compatibility and `dayjs.extend<T>()` as above, if a plugin ts definition doesn't pass a custom type for generic `T`, it fallbacks to `unknown`.

**Comment by iamkun** [comment]
THX

**Pull request merged**